### PR TITLE
[SC-298] Adjust README and more integration tests

### PR DESCRIPTION
**Pull request by hexonaut**
Adding integration tests to cover all the actual integrations that are expected on the mainnet market. Also, readme to explain the contracts.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -2,6 +2,26 @@
 
 This repository contains advanced features to improve SparkLend beyond the core contracts. This repository is mostly focused on improving security and automating governance processes.
 
+## Available Contracts
+
+### Oracles
+
+[FixedPriceOracle](https://github.com/marsfoundation/sparklend-advanced/blob/master/src/FixedPriceOracle.sol): A hardcoded oracle price that never changes. Uses: DAI market
```
Comment (lucas-manuel) [inline README.md]: Used for:

File: README.md [hunk 2]
```diff
@@ -2,6 +2,26 @@
...
+
+[CappedOracle](https://github.com/marsfoundation/sparklend-advanced/blob/master/src/CappedOracle.sol): Returns `min(market price, hardcoded max price)`. Uses: USDC/USDT markets
+
+### Custom Interest Rate Strategies
+
+[VariableBorrowInterestRateStrategy](https://github.com/marsfoundation/sparklend-advanced/blob/master/src/VariableBorrowInterestRateStrategy.sol): Modified version of `DefaultReserveInterestRateStrategy` that removes the stable borrow logic.
+
+[RateTargetBaseInterestRateStrategy](https://github.com/marsfoundation/sparklend-advanced/blob/master/src/RateTargetBaseInterestRateStrategy.sol): Overriden version of `VariableBorrowInterestRateStrategy` that sets the base variable borrow rate to match an external rate source with a fixed spread. Uses: DAI market
+
```
Comment (lucas-manuel) [inline README.md:18]: Overridden

File: README.md [hunk 3]
```diff
@@ -2,6 +2,26 @@
...
+[RateTargetKinkInterestRateStrategy](https://github.com/marsfoundation/sparklend-advanced/blob/master/src/RateTargetKinkInterestRateStrategy.sol): Overriden version of `VariableBorrowInterestRateStrategy` that sets the variable slope 1 rate to match an external rate source with a fixed spread. Uses: ETH/USDC/USDT markets
```
Comment (lucas-manuel) [inline README.md]: Overridden

Comment (lucas-manuel) [reply README.md]: Used for

File: README.md [hunk 4]
```diff
@@ -2,6 +2,26 @@
...
+
+### Rate Sources
+
+[PotRateSource](https://github.com/marsfoundation/sparklend-advanced/blob/master/src/PotRateSource.sol): Adapter to convert DSR into APR which can be consumed by one of the rate target interest rate strategies. Uses: DAI/USDC/USDT markets
```
Comment (lucas-manuel) [inline README.md]: Used for

File: test/SparkLendMainnetIntegration.t.sol [hunk 5]
```diff
@@ -0,0 +1,287 @@
...
+
+    // TODO add capped oracles for WBTC (need to import the combining contract first)
+
+    function _getBorrowRate(address asset) internal view returns (uint256) {
+        return pool.getReserveData(asset).currentVariableBorrowRate;
+    }
+
+    function _triggerUpdate(address asset) internal {
+        // Flashloan small amount to force indicies update
+        pool.flashLoanSimple(address(this), asset, 1, "", 0);
```
Comment (lucas-manuel) [inline test/SparkLendMainnetIntegration.t.sol:271]: indices

File: test/SparkLendMainnetIntegration.t.sol [hunk 6]
```diff
@@ -0,0 +1,287 @@
...
+        // Trigger an update from the new IRM
+        _triggerUpdate(DAI);
+
+        // Should be 0.5% higher than before
+        assertEq(_getBorrowRate(DAI), currentBorrowRate + 0.005e27);
+    }
+
+    function test_eth_market_irm() public {
+        // TODO replace with actual ETH yield oracle when ready
+        RateTargetKinkInterestRateStrategy strategy
```
Comment (lucas-manuel) [inline test/SparkLendMainnetIntegration.t.sol:127]:
```suggestion
        // TODO: Replace with actual ETH yield oracle when ready
```

File: test/SparkLendMainnetIntegration.t.sol [hunk 7]
```diff
@@ -0,0 +1,287 @@
...
+            = new RateTargetKinkInterestRateStrategy({
+                provider:                 poolAddressesProvider,
+                rateSource:               address(new RateSourceMock(0.03823984723902383709e27)),  // 3.8% (approx APR as of Dec 14, 2023)
```
Comment (lucas-manuel) [inline test/SparkLendMainnetIntegration.t.sol]: Lets assign this to a variable so we can use it in the assertions for easier reading

File: test/SparkLendMainnetIntegration.t.sol [hunk 8]
```diff
@@ -0,0 +1,287 @@
...
+
+        // slope1 has adjusted down a bit so the borrow rate is slightly lower at same utilization
+        assertEq(_getBorrowRate(ETH), 0.027800738894650680218981526e27);
+    }
+
+    function test_usdc_usdt_market_oracles() public {
+        CappedOracle usdcOracle = new CappedOracle(USDC_ORACLE, 1e8);
+        CappedOracle usdtOracle = new CappedOracle(USDT_ORACLE, 1e8);
+        
+        // Nothing is special about these numbers, they just happens to be the price at this block
```
Comment (lucas-manuel) [inline test/SparkLendMainnetIntegration.t.sol]: happen

File: test/SparkLendMainnetIntegration.t.sol [hunk 9]
```diff
@@ -0,0 +1,287 @@
...
+        vm.prank(ADMIN);
+        aaveOracle.setAssetSources(
+            assets,
+            sources
+        );
+
+        assertEq(aaveOracle.getAssetPrice(USDC), 1e8);
+        assertEq(aaveOracle.getAssetPrice(USDT), 0.99961441e8);
+
+        // TODO test if price of oracle drops below $1
```
Comment (lucas-manuel) [inline test/SparkLendMainnetIntegration.t.sol]: `// TODO: Test if price of oracle drops below $1`

Also what do you mean here, isn't this covered by USDT?

File: test/SparkLendMainnetIntegration.t.sol [hunk 10]
```diff
@@ -0,0 +1,287 @@
...
+
+        _triggerUpdate(USDC);
+        _triggerUpdate(USDT);
+
+        // Almost +0.2%, but utilization is slightly below the kink point
+        assertEq(_getBorrowRate(USDC), currentRateUSDC + 0.001891955013226013824859422e27, "after2: USDC mismatch");
+        assertEq(_getBorrowRate(USDT), currentRateUSDT + 0.001899234005225860815063204e27, "after2: USDT mismatch");
+    }
+
+    // TODO add capped oracles for WBTC (need to import the combining contract first)
```
Comment (lucas-manuel) [inline test/SparkLendMainnetIntegration.t.sol]: TODO: Add

Comment (lucas-manuel) [reply test/SparkLendMainnetIntegration.t.sol]: What does combining contract mean here?

File: test/SparkLendMainnetIntegration.t.sol [hunk 11]
```diff
@@ -0,0 +1,287 @@
...
+import { RateSourceMock } from "./mocks/RateSourceMock.sol";
+
+import { IAaveOracle }                  from "aave-v3-core/contracts/interfaces/IAaveOracle.sol";
+import { IPoolAddressesProvider }       from "aave-v3-core/contracts/interfaces/IPoolAddressesProvider.sol";
+import { IPool }                        from "aave-v3-core/contracts/interfaces/IPool.sol";
+import { IPoolConfigurator }            from "aave-v3-core/contracts/interfaces/IPoolConfigurator.sol";
+import { IDefaultInterestRateStrategy } from "aave-v3-core/contracts/interfaces/IDefaultInterestRateStrategy.sol";
+import { IERC20 }                       from "erc20-helpers/interfaces/IERC20.sol";
+import { SafeERC20 }                    from "erc20-helpers/SafeERC20.sol";
+
```
Comment (lucas-manuel) [inline test/SparkLendMainnetIntegration.t.sol]:
```suggestion
import { IAaveOracle }                  from "aave-v3-core/contracts/interfaces/IAaveOracle.sol";
import { IPoolAddressesProvider }       from "aave-v3-core/contracts/interfaces/IPoolAddressesProvider.sol";
import { IPool }                        from "aave-v3-core/contracts/interfaces/IPool.sol";
import { IPoolConfigurator }            from "aave-v3-core/contracts/interfaces/IPoolConfigurator.sol";
import { IDefaultInterestRateStrategy } from "aave-v3-core/contracts/interfaces/IDefaultInterestRateStrategy.sol";

import { IERC20 }    from "erc20-helpers/interfaces/IERC20.sol";
import { SafeERC20 } from "erc20-helpers/SafeERC20.sol";
```

File: test/SparkLendMainnetIntegration.t.sol [hunk 12]
```diff
@@ -0,0 +1,287 @@
...
+import { FixedPriceOracle }                   from "../src/FixedPriceOracle.sol";
+import { CappedOracle }                       from "../src/CappedOracle.sol";
+import { PotRateSource }                      from "../src/PotRateSource.sol";
+import { RateTargetBaseInterestRateStrategy } from "../src/RateTargetBaseInterestRateStrategy.sol";
+import { RateTargetKinkInterestRateStrategy } from "../src/RateTargetKinkInterestRateStrategy.sol";
+
```
Comment (lucas-manuel) [inline test/SparkLendMainnetIntegration.t.sol:16]:
```suggestion
import { FixedPriceOracle }                   from "src/FixedPriceOracle.sol";
import { CappedOracle }                       from "src/CappedOracle.sol";
import { PotRateSource }                      from "src/PotRateSource.sol";
import { RateTargetBaseInterestRateStrategy } from "src/RateTargetBaseInterestRateStrategy.sol";
import { RateTargetKinkInterestRateStrategy } from "src/RateTargetKinkInterestRateStrategy.sol";
```

**Review by lucas-manuel — CHANGES_REQUESTED** [review]

File: README.md (hunk 1 above)
Comment (hexonaut) [reply README.md]: fixed

File: README.md (hunk 2 above)
Comment (hexonaut) [reply README.md:18]: fixed

File: README.md (hunk 3 above)
Comment (hexonaut) [reply README.md]: fixed

File: README.md (hunk 4 above)
Comment (hexonaut) [reply README.md]: fixed

File: test/SparkLendMainnetIntegration.t.sol (hunk 5 above)
Comment (hexonaut) [reply test/SparkLendMainnetIntegration.t.sol:271]: fixed

File: test/SparkLendMainnetIntegration.t.sol (hunk 6 above)
Comment (hexonaut) [reply test/SparkLendMainnetIntegration.t.sol:127]: fixed

File: test/SparkLendMainnetIntegration.t.sol (hunk 8 above)
Comment (hexonaut) [reply test/SparkLendMainnetIntegration.t.sol]: fixed

File: test/SparkLendMainnetIntegration.t.sol (hunk 9 above)
Comment (hexonaut) [reply test/SparkLendMainnetIntegration.t.sol]: ah yes, I changed the block and forgot to remove the comment. fixed

File: test/SparkLendMainnetIntegration.t.sol (hunk 10 above)
Comment (hexonaut) [reply test/SparkLendMainnetIntegration.t.sol]: Fixed and moved comment to top of file like you normally do. WBTC oracle is a combination of WBTC/BTC * BTC/USD oracles. There is a contract that BGD made to combine these. We need to import that into this repo to support a capped WBTC/BTC price, but it needs to convert to USD.

File: test/SparkLendMainnetIntegration.t.sol (hunk 11 above)
Comment (hexonaut) [reply test/SparkLendMainnetIntegration.t.sol]: fixed

File: test/SparkLendMainnetIntegration.t.sol (hunk 12 above)
Comment (hexonaut) [reply test/SparkLendMainnetIntegration.t.sol:16]: fixed and ya I like this better ill stop using relative paths

File: test/SparkLendMainnetIntegration.t.sol (hunk 7 above)
Comment (hexonaut) [reply test/SparkLendMainnetIntegration.t.sol]: done

File: README.md [hunk 13]
```diff
@@ -2,6 +2,26 @@
...
+
+[VariableBorrowInterestRateStrategy](https://github.com/marsfoundation/sparklend-advanced/blob/master/src/VariableBorrowInterestRateStrategy.sol): Modified version of `DefaultReserveInterestRateStrategy` that removes the stable borrow logic.
+
+[RateTargetBaseInterestRateStrategy](https://github.com/marsfoundation/sparklend-advanced/blob/master/src/RateTargetBaseInterestRateStrategy.sol): Overridden version of `VariableBorrowInterestRateStrategy` that sets the base variable borrow rate to match an external rate source with a fixed spread. Used for: DAI market
+
+[RateTargetKinkInterestRateStrategy](https://github.com/marsfoundation/sparklend-advanced/blob/master/src/RateTargetKinkInterestRateStrategy.sol): Overridden version of `VariableBorrowInterestRateStrategy` that sets the variable slope 1 rate to match an external rate source with a fixed spread. Used for: ETH/USDC/USDT markets
+
+### Rate Sources
+
+[PotRateSource](https://github.com/marsfoundation/sparklend-advanced/blob/master/src/PotRateSource.sol): Adapter to convert DSR into APR which can be consumed by one of the rate target interest rate strategies. Uses: DAI/USDC/USDT markets
```
Comment (lucas-manuel) [inline README.md]: One more `Used for`

File: test/SparkLendMainnetIntegration.t.sol (hunk 6 above)
Comment (lucas-manuel) [reply test/SparkLendMainnetIntegration.t.sol:127]: `TODO:`

File: test/SparkLendMainnetIntegration.t.sol [hunk 14]
```diff
@@ -137,7 +140,7 @@ contract SparkLendMainnetIntegrationTest is Test {
 
         assertEq(strategy.getBaseVariableBorrowRate(),    prevStrategy.getBaseVariableBorrowRate());
         assertEq(prevStrategy.getVariableRateSlope1(),    0.032e27);
-        assertEq(strategy.getVariableRateSlope1(),        0.03023984723902383709e27);
+        assertEq(strategy.getVariableRateSlope1(),        mockETHYield - 0.008e27);
         assertEq(strategy.getVariableRateSlope2(),        prevStrategy.getVariableRateSlope2());
         assertEq(prevStrategy.getMaxVariableBorrowRate(), 1.232e27);
         assertEq(strategy.getMaxVariableBorrowRate(),     1.23023984723902383709e27);
```
Comment (lucas-manuel) [inline test/SparkLendMainnetIntegration.t.sol]: Use mockETHYield here as well

File: test/SparkLendMainnetIntegration.t.sol [hunk 15]
```diff
@@ -10,15 +10,17 @@ import { IPoolAddressesProvider }       from "aave-v3-core/contracts/interfaces/
...
+import { IERC20 }    from "erc20-helpers/interfaces/IERC20.sol";
+import { SafeERC20 } from "erc20-helpers/SafeERC20.sol";
 
+import { FixedPriceOracle }                   from "src/FixedPriceOracle.sol";
+import { CappedOracle }                       from "src/CappedOracle.sol";
+import { PotRateSource }                      from "src/PotRateSource.sol";
+import { RateTargetBaseInterestRateStrategy } from "src/RateTargetBaseInterestRateStrategy.sol";
+import { RateTargetKinkInterestRateStrategy } from "src/RateTargetKinkInterestRateStrategy.sol";
+
+// TODO Add capped oracles for WBTC (need to import the combining contract first)
```
Comment (lucas-manuel) [inline test/SparkLendMainnetIntegration.t.sol]: `TODO:`

**Review by lucas-manuel — CHANGES_REQUESTED** [review]

File: test/SparkLendMainnetIntegration.t.sol (hunk 6 above)
Comment (hexonaut) [reply test/SparkLendMainnetIntegration.t.sol:127]: fixed

File: test/SparkLendMainnetIntegration.t.sol (hunk 10 above)
Comment (hexonaut) [reply test/SparkLendMainnetIntegration.t.sol]: added colon :

File: README.md (hunk 13 above)
Comment (hexonaut) [reply README.md]: oops. fixed

File: test/SparkLendMainnetIntegration.t.sol (hunk 15 above)
Comment (hexonaut) [reply test/SparkLendMainnetIntegration.t.sol]: fixed

File: test/SparkLendMainnetIntegration.t.sol (hunk 14 above)
Comment (hexonaut) [reply test/SparkLendMainnetIntegration.t.sol]: fixed

**Review by lucas-manuel — APPROVED** [review]

**Review by barrutko — APPROVED** [review]

**Pull request merged**